File: node/src/core/store_postgres.js
// The Postgres store DRIVER — the same six-method interface as the SQLite driver, behind the store facade
// (store.js). Selected by store_factory when DATABASE_URL names Postgres (Supabase or any). `pg` (node-postgres) is
// an OPTIONAL dependency (optionalDependencies in package.json) imported LAZILY via dynamic import, so the default
// install stays zero-dep and the SQLite path never loads it.
//
// The dangerous translation (identical to the python/go drivers): SQLite's `do()` takes a whole-DB write lock
// with BEGIN IMMEDIATE *before* the read, serializing the insert-first/claim case (the exactly-once claim seam —
// single-use codes, the money path). Postgres has no IMMEDIATE, and `SELECT … FOR UPDATE` locks NOTHING on a
// not-yet-existing row — so it cannot serialize a claim. The correct analog is `pg_advisory_xact_lock(<key>)` held
// for the whole transaction: exactly one transaction holds the per-key lock at a time, so read-decide-write is
// serialized even when the row is absent. The lock key is PG's own `hashtextextended` over a LENGTH-PREFIXED
// `(ns,k)` encoding (64-bit, injective — distinct keys never collide).
//
// Values are stored as `text` — byte-identical to the SQLite/facade JSON, so the cross-language raw-value contract
// holds (jsonb would CANONICALIZE — reorder keys / strip whitespace — diverging the stored bytes ×backend). The
// `seq bigserial` gives `values()` a stable insertion order (SQLite's rowid has no PG equivalent).
//
// node specifics: the store surface is ASYNC, so this driver awaits pg naturally. A single pooled connection
// (`max: 1`) is the SAME single-connection model as go's SetMaxOpenConns(1) — and, with the synchronous `inDo`
// reentry guard (node is single-threaded and the do() callback is PURE+SYNC), a nested store call inside a do()
// callback throws at the guard BEFORE it can deadlock on the one held connection.

export function scrubDsn(url) {
  // Redact userinfo from a DSN for an error line — never echo the password. postgres://u:p@h → postgres://u:***@h.
  try {
    const i = url.indexOf('://');
    if (i < 0) return '<dsn>';
    const scheme = url.slice(0, i);
    const rest = url.slice(i + 3);
    const at = rest.indexOf('@');
    if (at < 0) return url;
    let user = rest.slice(0, at);
    const c = user.indexOf(':');
    if (c >= 0) user = user.slice(0, c);
    return `${scheme}://${user}:***@${rest.slice(at + 1)}`;
  } catch {
    return '<dsn>';
  }
}

// length-prefixed (ns,k) so the map to a lock string is INJECTIVE (distinct keys never share a lock); PG's
// hashtextextended turns it into the 64-bit advisory-lock key. A deployment picks ONE language, so node's
// (utf16-unit) length prefix only needs to be injective within node — and namespaces are ASCII constants.
function lockArg(ns, key) {
  return `${ns.length}:${ns}${key}`;
}

export function newPostgresDriver(url) {
  // SCRUB-ON-DELETE HONESTY (a real security-property difference, made LOUD not silent): SQLite's global
  // `PRAGMA secure_delete=ON` zeroes a freed row's bytes on every DELETE — secrets_vault DESTROY relies on it for a
  // true revocation. Postgres has NO row-level equivalent: a DELETE (and even an overwrite — MVCC keeps the old
  // tuple) leaves plaintext in dead heap tuples until VACUUM, and in any logical replica / backup. At-rest
  // encryption answers a DIFFERENT threat (disk theft), not a live read. Selecting Postgres SILENTLY downgrades the
  // scrub for EVERY domain — refuse to start until the operator ACKNOWLEDGES it. Crypto-shred is the real upgrade.
  // This check is SYNCHRONOUS and runs at getDriver()/import, so the app refuses to boot without the ack.
  if (!(process.env.SECURE_DELETE_ACK || '').trim()) {
    throw new Error(
      'the Postgres backend cannot scrub deleted bytes on delete (SQLite\'s secure_delete has no Postgres '
      + 'equivalent) — deleted data may persist in dead tuples until VACUUM and in replicas/backups. Set '
      + 'SECURE_DELETE_ACK=1 to acknowledge and proceed, or use the SQLite backend for the scrub guarantee.');
  }

  let pool = null;       // pg.Pool, created on first use (dynamic import keeps pg optional)
  let readyPromise = null;
  let inDo = false;      // reentry guard: a nested store call inside a do() callback throws (exact — node single-threaded, fn sync)

  function guard() {
    if (inDo) {
      throw new Error('store call inside a storeDo() callback: fn must be pure (no storeGet/storePut/storeValues/'
        + 'storeDelete/nextId/storeDo) — it gets the current value and returns the next');
    }
  }

  // ensureReady — memoized one-time: dynamic-import pg, build the single-connection pool, create the schema. Lazy
  // so the SQLite default never imports pg and a connect error surfaces loudly on first use (the ack check already
  // failed loud at construction). The CREATE TABLE statements tolerate the PG cold-start race (parity with python/go).
  function ensureReady() {
    if (!readyPromise) {
      readyPromise = (async () => {
        let pg;
        try {
          pg = (await import('pg')).default;
        } catch {
          throw new Error('DATABASE_URL names Postgres but the `pg` package is not installed — run '
            + '`npm install pg` (it is an optional dependency), or unset DATABASE_URL to use SQLite.');
        }
        pool = new pg.Pool({ connectionString: url, max: 1 });
        try {
          await initTable('CREATE TABLE IF NOT EXISTS _kv (ns text, k text, v text, seq bigserial, PRIMARY KEY (ns, k))');
          await initTable('CREATE TABLE IF NOT EXISTS _seq (name text PRIMARY KEY, n bigint)');
        } catch (e) {
          throw new Error(`could not connect to DATABASE_URL (${scrubDsn(url)}): ${(e && e.message) || e}`);
        }
      })();
    }
    return readyPromise;
  }

  async function initTable(ddl) {
    // `CREATE TABLE IF NOT EXISTS` is NOT atomic in Postgres: two cold-starting workers can both pass the existence
    // check then collide ("relation already exists" / "duplicate key" / "tuple concurrently updated"). The table DOES
    // exist after the race — treat it as success, not an error (parity with go's mustExecPG / python's _init_table).
    try {
      await pool.query(ddl);
    } catch (e) {
      const m = String((e && e.message) || e).toLowerCase();
      if (!m.includes('already exists') && !m.includes('duplicate key') && !m.includes('concurrently')) throw e;
    }
  }

  // RAW interface: get/values return raw json strings (or undefined); put/do take raw json; the facade marshals.
  return {
    get db() { return pool; }, // the raw backend handle (surface parity; white-box uses only; null until first use)

    async get(ns, key) {
      guard();
      await ensureReady();
      const r = await pool.query('SELECT v FROM _kv WHERE ns = $1 AND k = $2', [ns, key]);
      return r.rows.length ? r.rows[0].v : undefined;
    },

    async put(ns, key, raw) {
      guard();
      await ensureReady();
      await pool.query('INSERT INTO _kv (ns, k, v) VALUES ($1, $2, $3) ON CONFLICT (ns, k) DO UPDATE SET v = EXCLUDED.v',
        [ns, key, raw]);
    },

    async values(ns) {
      guard();
      await ensureReady();
      const r = await pool.query('SELECT v FROM _kv WHERE ns = $1 ORDER BY seq', [ns]);
      return r.rows.map((row) => row.v);
    },

    async delete(ns, key) {
      guard();
      await ensureReady();
      await pool.query('DELETE FROM _kv WHERE ns = $1 AND k = $2', [ns, key]);
    },

    async nextId(name) {
      guard();
      await ensureReady();
      const r = await pool.query(
        'INSERT INTO _seq (name, n) VALUES ($1, 1) ON CONFLICT (name) DO UPDATE SET n = _seq.n + 1 RETURNING n', [name]);
      return Number(r.rows[0].n);
    },

    // do — raw read-modify-write serialized across PROCESSES by the per-key transaction-scoped advisory lock (the
    // claim-safe analog of BEGIN IMMEDIATE). fn(rawCur | undefined) -> [rawNext | undefined, result]. fn is PURE+SYNC,
    // so the inDo flag is exact and the single held connection never deadlocks on a nested call (the guard throws first).
    async do(ns, key, fn) {
      guard();
      await ensureReady();
      const client = await pool.connect(); // the one pooled connection; concurrent do()s on it serialize (go's max-1 model)
      try {
        await client.query('BEGIN');
        // SOLE serializer of read-decide-write for the claim case (FOR UPDATE locks nothing on an absent row); held
        // for the whole transaction, so exactly one txn claims an absent key.
        await client.query('SELECT pg_advisory_xact_lock(hashtextextended($1, 0))', [lockArg(ns, key)]);
        const r = await client.query('SELECT v FROM _kv WHERE ns = $1 AND k = $2', [ns, key]);
        const rawCur = r.rows.length ? r.rows[0].v : undefined;
        inDo = true;
        let rawNext, result;
        try {
          [rawNext, result] = fn(rawCur); // PURE + SYNC: no await between inDo=true and inDo=false → one atomic block
        } finally {
          inDo = false;
        }
        if (rawNext !== undefined) {
          await client.query('INSERT INTO _kv (ns, k, v) VALUES ($1, $2, $3) ON CONFLICT (ns, k) DO UPDATE SET v = EXCLUDED.v',
            [ns, key, rawNext]);
        }
        await client.query('COMMIT');
        return result;
      } catch (err) {
        try { await client.query('ROLLBACK'); } catch { /* the transaction never opened */ }
        throw err;
      } finally {
        client.release();
      }
    },
  };
}

File: python/app_pkg/core/store_postgres.py
"""The Postgres store DRIVER — the same six-method interface as the SQLite driver, behind the store facade. Selected
by store_factory when DATABASE_URL names Postgres (Supabase or any). psycopg is an OPTIONAL dependency (a commented
line in requirements.txt) imported lazily, so the default install stays zero-dep.

The dangerous translation: SQLite's `do()` takes a whole-DB write lock with BEGIN IMMEDIATE *before* the read,
which serializes the insert-first/claim case (the exactly-once claim seam — single-use codes, the money path).
Postgres has no IMMEDIATE, and
`SELECT … FOR UPDATE` locks NOTHING on a not-yet-existing row — so it cannot serialize a claim. The correct analog is
`pg_advisory_xact_lock(<key>)` held for the whole transaction: exactly one transaction holds the per-key lock at a
time, so the read-decide-write is serialized even when the row is absent. The lock key is PG's own
`hashtextextended` over a LENGTH-PREFIXED `(ns,k)` encoding (64-bit, injective — distinct keys never collide via
ns‖k ambiguity, and identical across client languages since PG computes the hash).

Values are stored as `text` — byte-identical to the SQLite/​facade JSON, so the cross-language raw-value contract
holds (jsonb would CANONICALIZE — reorder keys / strip whitespace — diverging the stored bytes ×backend). The
`seq bigserial` gives `values()` a stable insertion order (SQLite's rowid has no PG equivalent). Track B (the
deferred query() pushdown) adds an expression index via `(v::jsonb)->>'field'` when a measured workload needs it."""
import psycopg

from .store_sqlite import StoreReentryError   # the SAME reentry exception ×backend (domains catch one class)


def scrub_dsn(url: str) -> str:
    """Redact userinfo from a DSN for an error/log line — never echo the password. postgres://u:p@h → postgres://u:***@h."""
    try:
        scheme, rest = url.split("://", 1)
        if "@" in rest:
            userinfo, host = rest.split("@", 1)
            user = userinfo.split(":", 1)[0]
            return f"{scheme}://{user}:***@{host}"
        return url
    except ValueError:
        return "<dsn>"


class PostgresDriver:
    """A single autocommit connection guarded by a process lock (the SQLite driver's model). Multi-worker safety:
    each worker process has its own connection; next_id is an atomic upsert-returning, do() is serialized across
    processes by a transaction-scoped advisory lock (pooler-safe — released at commit)."""

    def __init__(self, url: str):
        import os
        import threading
        # SCRUB-ON-DELETE HONESTY (a real security-property difference, made LOUD not silent): SQLite's global
        # `PRAGMA secure_delete=ON` zeroes a freed row's bytes on every DELETE — secrets_vault DESTROY relies on it
        # for a true revocation. Postgres has NO row-level equivalent: a DELETE (and even an overwrite — MVCC keeps
        # the old tuple) leaves the plaintext in dead heap tuples until VACUUM, and in any logical replica / backup.
        # At-rest encryption answers a DIFFERENT threat (disk theft), not a live read. So selecting Postgres SILENTLY
        # downgrades the scrub guarantee for EVERY domain — refuse to start until the operator ACKNOWLEDGES it. Set
        # SECURE_DELETE_ACK=1 to proceed (you accept deleted data may persist until VACUUM; rely on at-rest
        # encryption / crypto-shred). Crypto-shredding (a per-secret key destroyed on DESTROY) is the real upgrade.
        if not (os.environ.get("SECURE_DELETE_ACK") or "").strip():
            raise RuntimeError(
                "the Postgres backend cannot scrub deleted bytes on delete (SQLite's secure_delete has no Postgres "
                "equivalent) — deleted data may persist in dead tuples until VACUUM and in replicas/backups. Set "
                "SECURE_DELETE_ACK=1 to acknowledge and proceed, or use the SQLite backend for the scrub guarantee.")
        try:
            self._conn = psycopg.connect(url, autocommit=True)
        except psycopg.Error as e:
            raise RuntimeError(f"could not connect to DATABASE_URL ({scrub_dsn(url)}): {e.__class__.__name__}") from None
        self._lock = threading.Lock()
        self._local = threading.local()
        with self._lock:
            self._init_table("CREATE TABLE IF NOT EXISTS _kv (ns text, k text, v text, seq bigserial, PRIMARY KEY (ns, k))")
            self._init_table("CREATE TABLE IF NOT EXISTS _seq (name text PRIMARY KEY, n bigint)")

    def _init_table(self, ddl: str) -> None:
        # `CREATE TABLE IF NOT EXISTS` is NOT atomic in Postgres: two cold-starting workers can both pass the
        # existence check then collide ("relation already exists" / "duplicate key" / "tuple concurrently updated").
        # The table DOES exist after the race — treat it as success, not an error (parity with go's mustExecPG).
        try:
            self._conn.execute(ddl)
        except psycopg.Error as e:
            msg = str(e).lower()
            if "already exists" not in msg and "duplicate key" not in msg and "concurrently" not in msg:
                raise

    def _guard(self) -> None:
        if getattr(self._local, "in_do", False):
            raise StoreReentryError(
                "store call inside a do() callback: fn must be pure (no get/put/values/delete/next_id/do) — it gets "
                "the current value and returns the next")

    @staticmethod
    def _lock_arg(ns: str, key: str) -> str:
        # length-prefixed so the (ns,k) -> string map is INJECTIVE (distinct keys never share a lock); PG's
        # hashtextextended turns it into the 64-bit advisory-lock key, identically across client languages.
        return f"{len(ns)}:{ns}{key}"

    def get(self, ns: str, key: str):
        self._guard()
        with self._lock:
            row = self._conn.execute("SELECT v FROM _kv WHERE ns = %s AND k = %s", (ns, key)).fetchone()
        return row[0] if row else None

    def put(self, ns: str, key: str, raw: str) -> None:
        self._guard()
        with self._lock:
            self._conn.execute("INSERT INTO _kv (ns, k, v) VALUES (%s, %s, %s) "
                               "ON CONFLICT (ns, k) DO UPDATE SET v = EXCLUDED.v", (ns, key, raw))

    def values(self, ns: str) -> list:
        self._guard()
        with self._lock:
            rows = self._conn.execute("SELECT v FROM _kv WHERE ns = %s ORDER BY seq", (ns,)).fetchall()
        return [r[0] for r in rows]

    def delete(self, ns: str, key: str) -> None:
        self._guard()
        with self._lock:
            self._conn.execute("DELETE FROM _kv WHERE ns = %s AND k = %s", (ns, key))

    def next_id(self, name: str) -> int:
        self._guard()
        with self._lock:
            return self._conn.execute(
                "INSERT INTO _seq (name, n) VALUES (%s, 1) "
                "ON CONFLICT (name) DO UPDATE SET n = _seq.n + 1 RETURNING n", (name,)).fetchone()[0]

    def do(self, ns: str, key: str, raw_fn):
        """raw_fn(raw_current | None) -> (raw_next | None, ret). Serialized across processes by the per-key
        transaction-scoped advisory lock (the claim-safe analog of BEGIN IMMEDIATE)."""
        self._guard()
        with self._lock:
            with self._conn.transaction():
                self._conn.execute("SELECT pg_advisory_xact_lock(hashtextextended(%s, 0))", (self._lock_arg(ns, key),))
                row = self._conn.execute("SELECT v FROM _kv WHERE ns = %s AND k = %s", (ns, key)).fetchone()
                raw_current = row[0] if row else None
                self._local.in_do = True
                try:
                    raw_next, ret = raw_fn(raw_current)
                finally:
                    self._local.in_do = False
                if raw_next is not None:
                    self._conn.execute("INSERT INTO _kv (ns, k, v) VALUES (%s, %s, %s) "
                                       "ON CONFLICT (ns, k) DO UPDATE SET v = EXCLUDED.v", (ns, key, raw_next))
            return ret
